src/server/auth/server/api_server.go

```diff
@@ -1416,22 +1366,163 @@ func (a *apiServer) RestoreAuthToken(ctx context.Context, req *auth.RestoreAuthT
 		}
 	}
 
-	// Check whether the token hash already exists - we don't want to replace an existing token
-	if _, err := col.NewSTM(ctx, a.env.GetEtcdClient(), func(stm col.STM) error {
-		tokens := a.tokens.ReadWrite(stm)
-		var existing auth.TokenInfo
-		err := tokens.Get(req.Token.HashedToken, &existing)
-		if err == nil {
-			return errors.New("cannot overwrite existing token with same hash")
-		} else if err != nil && !col.IsErrNotFound(err) {
-			return err
+	if err := func() error {
+		if ttl > 0 {
+			return a.insertAuthToken(ctx, req.Token.HashedToken, req.Token.TokenInfo.Subject, ttl)
+		} else {
+			return a.insertAuthTokenNoTTL(ctx, req.Token.HashedToken, req.Token.TokenInfo.Subject)
 		}
-
-		return tokens.PutTTL(req.Token.HashedToken,
-			req.Token.TokenInfo,
-			ttl)
-	}); err != nil {
+	}(); err != nil {
 		return nil, errors.Wrapf(err, "error restoring auth token")
 	}
+
 	return &auth.RestoreAuthTokenResponse{}, nil
 }
+
+// implements the protobuf auth.DeleteExpiredAuthTokens RPC
+func (a *apiServer) DeleteExpiredAuthTokens(ctx context.Context, req *auth.DeleteExpiredAuthTokensRequest) (*auth.DeleteExpiredAuthTokensResponse, error) {
+	logrus.Info("Deleting expired auth tokens")
+	if _, err := a.env.GetDBClient().Exec(`DELETE FROM auth.auth_tokens WHERE NOW() > expiration`); err != nil {
+		return nil, errors.Wrapf(err, "error deleting expired tokens")
+	}
+	return &auth.DeleteExpiredAuthTokensResponse{}, nil
+}
+
+func (a *apiServer) deleteExpiredTokensRoutine() {
+	go func(ctx context.Context) {
+		for {
+			time.Sleep(time.Duration(cleanupIntervalHours) * time.Hour)
+			a.DeleteExpiredAuthTokens(ctx, &auth.DeleteExpiredAuthTokensRequest{})
+		}
+	}(context.Background())
+}
+
+// we interpret an expiration value of NULL as "lives forever".
+// Now we will sometimes have expiration values set in the passed, since we only remove those values in the deleteExpiredTokensRoutine() goroutine
+func (a *apiServer) lookupAuthTokenExpiration(ctx context.Context, tokenHash string) (*types.Timestamp, error) {
+	var expiration *time.Time
+	if err := a.env.GetDBClient().GetContext(ctx, &expiration, `SELECT expiration FROM auth.auth_tokens WHERE token_hash = $1`, tokenHash); err != nil {
+		if err == sql.ErrNoRows {
+			return nil, col.ErrNotFound{Type: "auth_tokens", Key: tokenHash}
+		}
+		return nil, err
+	}
+
+	if expiration != nil {
+		expiryTS, tsErr := types.TimestampProto(*expiration)
+		if tsErr != nil {
+			return nil, tsErr
+		}
+		return expiryTS, nil
+	}
+	return nil, nil
+}
+
+func (a *apiServer) lookupAuthTokenInfo(ctx context.Context, tokenHash string) (*auth.TokenInfo, error) {
+	var tokenInfo auth.TokenInfo
+	if err := a.env.GetDBClient().GetContext(ctx, &tokenInfo, `SELECT subject FROM auth.auth_tokens WHERE token_hash = $1`, tokenHash); err != nil {
```

Comment:
I think this is missing the case where a token is expired but it hasn't been deleted yet? It might be good to add that to the test, call `WhoAmI` or something with the recently-expired-but-undeleted token and make sure it isn't accepted.